Dev 1.0 - Version 0.7

cgosingtian opened the pull request:
70% complete; remaining: Spawn enemy ships, implement collision, air bomb.
Optional target: ground enemies, ground bullets

----- Object Details and Relationships -----
[Controllers]
Battle View Controller
- main view controller, sets up game elements in a XIB file.
  - creates 1 Player Ship View
  - creates 4 Cloud Views
  - creates 1 Background View
- Instantiates Player Controller
  - For each Turret in Player's Ship, instantiate Turret Controllers
- Has Touch Control Recognition (Long Press and Swipe)
  - Delegates control to Player Controller

Player Controller
- Instantiates Player Ship and Player
- Performs Player Ship movement (touch control comes Battle View Controller)
  - Instructs Player Ship View to animate when necessary

Turret Controller
- Creates Bullet Controllers based on Turret data (firing speed, bullet type, etc)
  - Repeatedly creates and releases Bullet Controllers per Bullet
- Ensures that Bullet Controllers are fired from the Player Ship that owns the Turret

Bullet Controllers
- Creates a Bullet View from a Bullet
- Handles the Bullet's movement
- Plays a sound when a Bullet is fired

[Views]
Ship View
- Contains method for moving View frame

Player Ship View : Ship View
- Animates movement of Player Ship

Bullet View
- Displays and animates Bullets created by Bullet Controller

[Models]
Ship
- Owns a mutable array of Turrets
- Contains health, air bombs, velocity, coordinates, points awarded (if enemy), "is player" flag

Turret
- Weak reference to Ship (that owns it) and Bullet
- Contains turret type, firing speed and angle data

Bullet
- Contains velocity, damage, "is air bullet" flag, "is player bullet" flag, bullet type

Player
- Contains score and name

Miscellaneous Views:
Cloud View
- UIImageView with cloud graphic, set up in Battle View Controller's XIB
- Core animation translates it downward, fading in and out over its lifetime
- Respawns at a random location and with a random size (width is always 2.5x height) after animation

Background View
- UIImageView with floor graphic, set up in Battle View Controller's XIB
- Manual animation through NSTimer and a repeating selector
- Resets its location before the whole image is traversed to create a fake "unlimited floor" effect

----- EXERCISE TOPICS ADDRESSED -----
- Non-ARC Memory Management
- Animations
  - View Animation: (KLBCloudView, KLBBackgroundView)
  - UIImageView Animation: (KLBBulletView)
  - Core Animation: (KLBPlayerShipView)
- Custom Font: (KLBBattleViewController.xib)
- Audio (AVAudioPlayer): (KLBBulletController)
- MVC
- Blocks: (KLBBulletView, KLBCloudView - mostly in animations)

Commit:
Created Model object files for Ship, Turret, and Bullet. Added basic properties for these objects:

Ship
- Health Maximum (int)
- Health Current (int)
- Air Bombs (int)
- Velocity (CGFloat)
- X coordinate (CGFloat)
- Y coordinate (CGFloat)
- Turrets

Turret
- Angle (CGFloat)
- Firing Speed (CGFloat)
- Bullet

Bullet
- Speed (CGFloat)
- Damage (int)
- Air/Ground (bool)

Commit:
Implemented initialisers and key methods for Bullet, Turret, and Ship.

Commit:
CHANGELOG-8192014-1600
MODELS:
Added Player model object
Added score property to Ship
VIEWS:
Added Ship View
Added Player Ship View
Added basic neutral, player, and enemy ship graphics
CONTROLLERS:
Added Player Controller
Added touch controls to Battle View Controller
 - basic left and right movement
 - placeholder for bomb launching via swipe forward
PROTOCOLS:
Added Ship Control Protocol for movement and firing

Commit:
CHANGELOG-8192014-1715
VIEWS:
Added bullet graphics for bullet types default, machine-gun, cannon
CONTROLLERS:
Changed touch controls to use gesture recognisers
 - long press creates repeating timer for continuous movement
 - pan vertically to throw a bomb
 - Controls are very rudimentary

Commit:
CHANGELOG-8202014-1136
MODELS:
Added getters to Ship and Turret for use in their Controllers
Updated Turret to have reference to Ship Owner
 - updated Ship to use the correct initialiser
VIEWS:
Created Bullet View to display bullets
CONTROLLERS:
Created Turret Controller that handles the firing of Bullets
 - Creates a Bullet Controller every interval based on Turret firing speed
Created Bullet Controller that fires a bullet in one direction
Added movement limits to Player Controller to prevent moving past window
PROTOCOLS:
Added shipDidMove method to Ship Control Protocol

Commit:
CHANGELOG-820-2014-1202
VIEWS:
Added simple animation for Ship turning right/left
 - Used 3D Rotation/Transformation on a UIImageView's CALayer

Commit:
CHANGELOG-820-2014-1404
VIEWS:
Added floor graphic and animation to Battle XIB
Added another image per bullet type for animation
Added bullet animations
 - keyframe fade in from invisible
 - keyframe switch between alternate bullet image

Commit:
CHANGELOG-820-2014-1422
VIEWS:
Fixed visual bug that caused player ship's layer to overlap other layers
 - set player ship view layer's zPosition to 100
Added Cloud image and view
 - Used Core Animation to animate: fade in/out and translate down over time

Commit:
CHANGELOG-820-2014-1630
Added custom font "OCR A Std"
Changed all fonts of labels to "OCR A Std"
Added laser firing sound, played on bullet launch

Commit:
CHANGELOG-820-2014-1638
VIEWS:
Converted laser.wav to IMA4 compression, MONO
 - with afconvert -f AIFC -d ima4 -c 1 laser.wav
Made the background animation smoother

Commit:
CHANGELOG-820-2014-1739
Fixed memory leaks
 - (Bullet firing still creates memory leaks)

Commit:
CHANGELOG-820-2014-1825
Fixed bullet-firing memory leak
 - 16 byte static memory leak with AVAudioPlayer remains
Cleaned up comments and debug code

Commit:
CHANGELOG-821-2014-0854
Changed some class names to fit their MVC role more accurately
- Changed KLBClouds to KLBCloudView
- Changed KLBBackground to KLBBackgroundView
Added README file

Commit:
Fixed bug in initialiser of Bullet
- "super init" instead of "self init"

Commit:
Merge branch 'master' into dev_1.0
- Added README.md to dev_1.0 branch

Commit:
Updated README.md

cgosingtian committed:
Update README.md

Updated README.md to match commentary from latest pull request.

Commit:
Moved images, sound effects, and fonts to Resources folder.

ericacaber commented on Air Raid/KLBBackgroundView.m:
```diff
@@ -0,0 +1,69 @@
+@synthesize maxHeight = _maxHeight,
+            distanceTravelled = _distanceTravelled,
+            moveSpeed = _moveSpeed,
+            animationInterval = _animationInterval;
```
You can remove this po. @synthesize is automatically called when you create a property.
https://developer.apple.com/library/mac/releasenotes/DeveloperTools/RN-Xcode/xc4_release_notes/xc4_release_notes.html#//apple_ref/doc/uid/TP40001051-CH3-SW145

cgosingtian replied:
Removed all synthesize calls. Thanks.

ericacaber commented on Air Raid/KLBBackgroundView.m:
```diff
@@ -0,0 +1,69 @@
+#pragma mark - View States
+- (void)awakeFromNib {
+    _maxHeight = self.image.size.height;
+    NSTimer *timer = [NSTimer scheduledTimerWithTimeInterval:_animationInterval target:self selector:@selector(animateBackground) userInfo:nil repeats:YES];
```
You can also do this declaration like this
![screen shot 2014-08-26 at 10 37 03 am](https://cloud.githubusercontent.com/assets/2919905/4039353/efee7868-2cc9-11e4-9d9c-6e195b4d1e49.png)
it is more readable than declaring it on one line.

rogerKlab replied:
in this implementation timer is unused variable, you should add "timer" in runloops
https://developer.apple.com/library/mac/documentation/Cocoa/Reference/Foundation/Classes/NSTimer_Class/Reference/NSTimer.html

cgosingtian replied:
Thanks, will fix readability. It does look a bit better that way.

With regards to timer being unused - I used scheduledTimerWithTimeInterval:target:selector:userInfo:repeats: to "create the timer and schedule it on the current run loop in the default mode." I didn't use it again within the same method so the compiler warned me, but I'm sure it works fine. I invalidated it explicitly within the same file.

ericacaber commented on Air Raid/KLBBattleViewController.m:
```diff
@@ -0,0 +1,168 @@
+                                                                 action:@selector(handleLongPress:)];
+        _longPressRecognizer.delegate = self;
+        _longPressRecognizer.minimumPressDuration = 0.05;
+        _longPressRecognizer.allowableMovement = 600.0;
```
Avoid using magic numbers po. you can use const for this.

cgosingtian replied:
Removed all magic numbers and replaced them with constants.

ericacaber commented on Air Raid/KLBBullet.m:
```diff
@@ -0,0 +1,61 @@
+@synthesize velocity = _velocity,
+            damage = _damage,
+            isAirBullet = _isAirBullet,
+            bulletType = _bulletType;
```
You can remove this po. @synthesize is automatically called when you create a property.
https://developer.apple.com/library/mac/releasenotes/DeveloperTools/RN-Xcode/xc4_release_notes/xc4_release_notes.html#//apple_ref/doc/uid/TP40001051-CH3-SW145

cgosingtian replied:
Removed all synthesize calls. Thanks.

ericacaber commented on Air Raid/KLBBullet.m:
```diff
@@ -0,0 +1,61 @@
+- (instancetype) initWithBulletType:(BulletTypes)b playerBullet:(bool)p {
+    self = [super init];
+    _playerBullet = p;
+    switch (b) {
```
What is b? Use more descriptive variable name po.

cgosingtian replied:
Changing "b" to "bulletType" and "p" to "isPlayerBullet" for clarity since the initializer uses them quite a bit. Will fix other instances of this accordingly. Thanks!

ericacaber commented on Air Raid/KLBBullet.m:
```diff
@@ -0,0 +1,61 @@
+    _bulletType = b;
+    _velocity = v;
+    _damage = d;
+    _isAirBullet = a;
```
Use more descriptive variable name po.

cgosingtian replied:
Done.

ericacaber commented on Air Raid/KLBBulletController.m:
```diff
@@ -0,0 +1,178 @@
+            coordinates = _coordinates,
+            traversedDistance = _traversedDistance,
+            timer = _timer,
+            avPlayer = _avPlayer;
```
You can remove this po. @synthesize is automatically called when you create a property.
https://developer.apple.com/library/mac/releasenotes/DeveloperTools/RN-Xcode/xc4_release_notes/xc4_release_notes.html#//apple_ref/doc/uid/TP40001051-CH3-SW145

cgosingtian replied:
Removed all synthesize calls. Thanks.

ericacaber commented on Air Raid/KLBBulletController.m:
```diff
@@ -0,0 +1,178 @@
+    
+    [self setTimer:nil];
+//    _bulletView = nil;
+//    _bullet = nil;
```
If the comment out code is not needed anymore. you can erase it.

cgosingtian replied:
I'll keep the commented code to warn other coders not to try them (the other classes take care of deallocating its parts). I'll leave a note explaining why in the code.

ericacaber commented on Air Raid/KLBBulletController.m:
```diff
@@ -0,0 +1,178 @@
+        [bv release];
+        _bullet = b;
+        _launchAngle = [self limitAngle:a];
+        _coordinates = c;
```
Use more descriptive variable name po. avoid using letters or acronym. it's not readable anymore.

cgosingtian replied:
Done - it does look better.

ericacaber commented on Air Raid/KLBBulletView.m:
```diff
@@ -0,0 +1,157 @@
+    else {
+        UIImage *bulletImage = [UIImage imageNamed:@"bullet_default.png"];
+        self.image = bulletImage;
+    }
```
you can use switch here po.

cgosingtian replied:
Used a switch as recommended.

ericacaber commented on Air Raid/KLBBulletView.m:
```diff
@@ -0,0 +1,157 @@
+                else {
+                    UIImage *bulletImage = [UIImage imageNamed:@"bullet_default2.png"];
+                    self.image = bulletImage;
+                }
```
you can use switch here po

cgosingtian replied:
Used a switch as recommended.

ericacaber commented on Air Raid/KLBPlayer.m:
```diff
@@ -0,0 +1,31 @@
+@implementation KLBPlayer
+
+@synthesize score = _score,
+            name = _name;
```
You can remove this po. @synthesize is automatically called when you create a property.
https://developer.apple.com/library/mac/releasenotes/DeveloperTools/RN-Xcode/xc4_release_notes/xc4_release_notes.html#//apple_ref/doc/uid/TP40001051-CH3-SW145

cgosingtian replied:
Removed all synthesize calls. Thanks.

ericacaber commented on Air Raid/KLBTurret.m:
```diff
@@ -0,0 +1,99 @@
+            firingSpeed = _firingSpeed,
+            bullet = _bullet,
+            turretType = _turretType,
+            owner = _owner;
```
You can remove this po. @synthesize is automatically called when you create a property.
https://developer.apple.com/library/mac/releasenotes/DeveloperTools/RN-Xcode/xc4_release_notes/xc4_release_notes.html#//apple_ref/doc/uid/TP40001051-CH3-SW145

cgosingtian replied:
Removed all synthesize calls. Thanks.

ericacaber commented on Air Raid/KLBTurretController.m:
```diff
@@ -0,0 +1,65 @@
+    NSLog(@"Deallocating Turret");
+    [_timer invalidate];
+    [_timer release];
+    [_turret release];
```
turret was declared retain. If you declare a property with retain semantics, you can use _turret = nil; on releasing it, you can already remove [_turret release]. 
http://stackoverflow.com/questions/4220815/release-vs-nil-best-practice

cgosingtian replied:
Good call, but wouldn't we want to ensure safety in case the setter for turret is changed in the future by another coder? It's still the recommended option by the top-marked answer to release and then nil, and so far I've not encountered any problems with doing both. Of course, I will defer to your experience if you still think otherwise.

ericacaber commented on Air Raid/KLBBackgroundView.m:
```diff
@@ -0,0 +1,69 @@
+}
+
+- (void) resetPosition {
+    [self setFrame:CGRectMake(0.0, -(self.image.size.height/2), self.image.size.width, self.image.size.height)];
```
0.0 was used many times. You can use const. Avoid using magic numbers po.

rogerKlab replied:
image : floor.png size (320x960)

in 3.5 device 320x480 like iphone4
-(self.image.size.height/2) = 480
 it will fit in the 480 screen

but in 4.0 device 320x568 like iphone5
-(self.image.size.height/2) = 480 and screen size is 568
image will not fit in the screen,

![ios simulator screen shot aug 26 2014 4 39 20 pm](https://cloud.githubusercontent.com/assets/2919821/4041892/b023307e-2cfc-11e4-9746-aaa2b8ae3eae.png)

its always advisable to refer in screen size when positioning object in the view
you can use [[UIScreen mainScreen]bounds] to get the screen size or CGGeometry "CGRectGetHeight" to get the view height

cgosingtian replied:
Fixed above issues - good tip on screen bounds too.

Commit:
CHANGELOG-826-2014-1155
MODELS:
Fixed missing initialisations in default Ship init
VIEWS:
Added initialiser with coordinates for Ship View
Added animations for Ship View
CONTROLLERS:
Added EnemyShipController
Implemented Bomb notification posting in Player Controller

ericacaber commented on Air Raid/KLBBattleViewController.m:
```diff
@@ -0,0 +1,168 @@
+            playerShipView = _playerShipView,
+            playerControllerDelegate = _playerControllerDelegate,
+            longPressRecognizer = _longPressRecognizer,
+            panRecognizer = _panRecognizer;
```
You can remove this po. @synthesize is automatically called when you create a property.
https://developer.apple.com/library/mac/releasenotes/DeveloperTools/RN-Xcode/xc4_release_notes/xc4_release_notes.html#//apple_ref/doc/uid/TP40001051-CH3-SW145

cgosingtian replied:
Removed them all, thanks.

Commit:
Fixed boolean naming, standardised order of property attributes in declarations, reactivated player ship movement.

ericacaber commented on Air Raid/KLBBattleViewController.m:
```diff
@@ -0,0 +1,168 @@
+    _panRecognizer = nil;
+    
+    [_scorePlaceholderLabel release];
+    [_scoreActualLabel release];
```
If you declare a property with retain semantics, you can use self.variable = nil; on releasing it, you can already remove [_variable release]. 

_variable = nil; is the same as [_variable release], _variable= nil;

http://stackoverflow.com/questions/4220815/release-vs-nil-best-practice

cgosingtian replied:
Going to keep the release-nil syntax for now for safety. In the future I might just follow Mr. RegiePinat's suggestion to use the setter (self.blah = nil) to do both. Thanks!

ericacaber commented on Air Raid/KLBBattleViewController.m:
```diff
@@ -0,0 +1,168 @@
+        _panRecognizer.delegate = self;
+        [self.view addGestureRecognizer:_panRecognizer];
+        
+        _scorePlaceholderLabel.font = [UIFont fontWithName:@"OCR A Std" size:KLB_FONT_SIZE_STANDARD];
```
you can also use string const on font name

cgosingtian replied:
Did this. Removed all magic numbers too.

ericacaber commented on Air Raid/KLBBullet.h:
```diff
@@ -0,0 +1,30 @@
+                //Damage Velocity isAirBullet
+    btDefault,    //  2    10.0        y
+    btMachineGun, //  1    20.0        y
+    btCannon,     //  4     5.0        y
```
remove unnecessary comments po

cgosingtian replied:
Removed them all.

ericacaber commented on Air Raid/KLBBullet.h:
```diff
@@ -0,0 +1,30 @@
+@property (nonatomic) bool playerBullet;
+
+- (instancetype) initWithBulletType:(BulletTypes)b playerBullet:(bool)p;
+- (bool) playerBullet;
```
It is more appropriate to use BOOL rather than bool
http://stackoverflow.com/questions/541289/objective-c-bool-vs-bool

cgosingtian replied:
Done. Changed all bools to BOOL.

rogerKlab commented on Air Raid/KLBBattleViewController.m:
```diff
@@ -0,0 +1,168 @@
+
+@property (nonatomic) bool hasMovedTouch;
+@property (nonatomic) CGPoint touchedLocation;
+@property (retain, nonatomic) KLBPlayerController<KLBShipControlProtocol> *playerControllerDelegate;
```
In this case "KLBEnemyShipController" retain the object "KLBPlayerController" and create a strong reference at same time "KLBPlayerController" retain the object it self.

"playerControllerDelegate" property attribute should be "assign" to avoid retain cycles 
Retain cycles:
http://www.cocoawithlove.com/2009/07/rules-to-avoid-retain-cycles.html

Retaining objects create a strong reference the object created cannot be deallocated until all of its strong reference are release.

cgosingtian replied:
Renamed playerControllerDelegate to playerController to make it clearer. It doesn't have a reference to its "delegatee" so there's no retain cycle. Verified this in the Leaks Instruments.

rogerKlab commented on Air Raid/KLBBattleViewController.m:
```diff
@@ -0,0 +1,168 @@
+    _longPressRecognizer.delegate = nil;
+    _panRecognizer.delegate = nil;
+    
+    [_playerControllerDelegate release];
```
"playerControllerDelegate" should not release, it will cause EXC_BAD_ACCESS when object in delegate class deallocated, "playerControllerDelegate" doesn't owned by "KLBBattleViewController", and delegate should always release by its parent class.

cgosingtian replied:
Renamed playerControllerDelegate to playerController to avoid confusion. I suppose I wasn't exactly using it as a means to inform my main view controller that playerController finished doing something - it was just a means for me to ensure that my playerController implemented certain methods.

rogerKlab commented on Air Raid/KLBBullet.m:
```diff
@@ -0,0 +1,61 @@
+}
+
+#pragma mark - Getters and Setters
+- (void) setBulletType:(BulletTypes)b velocity:(CGFloat)v damage:(int)d isAirBullet:(bool)a {
```
please use NSInteger or NSUInteger data type for parameters to maintain binary compatibility,
in 32-bit NSInterger and NSUInteger represent "int" but in 64-bit architecture like iPhone5s  NSInterger and NSUInteger represent "long". Using "int" in control flow like for loops iteration well be fine, but in passing values between two objects via its method its advisable to use Cocoa API' like NSInteger in representing integer.
As well as "float" to CGFloat

https://developer.apple.com/library/mac/documentation/Cocoa/Conceptual/Cocoa64BitGuide/64BitChangesCocoa/64BitChangesCocoa.html

cgosingtian replied:
Done - changed all floats to CGFloats and ints to NSUInteger. Thanks!

rogerKlab commented on Air Raid/KLBConstants.m:
```diff
@@ -0,0 +1,17 @@
+//
+//  KLBConstants.m
+//  Air Raid
```
KLBConstants class only used for declaring constant value, implementation body is not needed,
When declaring constant you can use the native "C" header or .h file

cgosingtian replied:
I designed the constants file to be for globals (eg. font strings, font size, animation speed), though I did keep class-specific constants in their own files.

rogerKlab commented on Air Raid/KLBPlayerShipView.h:
```diff
@@ -0,0 +1,20 @@
+#import <UIKit/UIKit.h>
+#import "KLBShipView.h"
+#import "KLBShipControlProtocol.h"
+@class KLBShip;
```
unused declaration, no "KLBShip" objects or method parameter found in .h file

rogerKlab commented on Air Raid/KLBShipView.h:
```diff
@@ -0,0 +1,17 @@
+
+#import <UIKit/UIKit.h>
+@class KLBShip;
+
```
unused declaration, no "KLBShip" objects or method parameter found in .h file

cgosingtian replied:
Took this out. Thanks.

rogerKlab commented on Air Raid/KLBPlayerController.m:
```diff
@@ -0,0 +1,141 @@
+-(void)shipDidLaunchBomb {
+    NSLog(@"SHIP DID LAUNCH BOMB");
+    [[NSNotificationCenter defaultCenter] postNotificationName:KLB_BOMB_EXPLOSION_NOTIFICATION object:self];
+}
```
No observer found for "KLB_BOMB_EXPLOSION_NOTIFICATION" notification key

rogerKlab commented on Air Raid/KLBBattleViewController.m:
```diff
@@ -0,0 +1,168 @@
+//  Created by Chase Gosingtian on 8/19/14.
+//  Copyright (c) 2014 KLab Cyscorpions, Inc. All rights reserved.
+//
+
```
Please disable device autorotate when orientation is not supported.

![ios simulator screen shot aug 26 2014 3 57 14 pm](https://cloud.githubusercontent.com/assets/2919821/4041617/4c71aa82-2cf8-11e4-9e67-48fee74a4132.png)

but if you wish to support multiple device orientation you can use the UIViewController method "shouldAutorotateToInterfaceOrientation" to handle orientations.

cgosingtian replied:
Good catch - will disable autorotate.

Commit:
CHANGELOG-826-2014-1628
Removed explicit synthesise calls.
Added line breaks for messages with multiple arguments.
Removed magic numbers - added them to Constants (global) or to files locally.
Made some variable names more verbose.
Bullet: Renamed "initWithBulletType" to "initWithBullet" for clarity.
Bullet: Renamed "btDefault" to "btDefaultBulletType" for clarity.
Changed all bools to BOOLs, and true/false to YES/NO accordingly.
Removed some comments.

Commit:
With test labels for the accelerometer.

Commit:
CHANGELOG-827-2014-1204
Removed all magic numbers.
 - Constants placed in their respective classes; global constants in Constants file
Changed all ints to NSUInteger.
Changed all floats to CGFloat.
Changed object positioning to follow Screen bounds.
 - Fixed bug with Background View not scaling to the screen properly
Fixed "snapping" of Cloud View between animation cycles

Commit:
Disabled auto-rotate.

cgosingtian commented:
I think I've managed to address all issues - summarized changelog:

Removed explicit synthesise calls.
Added line breaks for messages with multiple arguments.
Made some variable names more verbose.
Bullet: Renamed "initWithBulletType" to "initWithBullet" for clarity.
Bullet: Renamed "btDefault" to "btDefaultBulletType" for clarity.
Changed all bools to BOOLs, and true/false to YES/NO accordingly.
Removed some comments.
Removed all magic numbers.
- Constants placed in their respective classes; global constants in Constants file

Changed all ints to NSUInteger.
Changed all floats to CGFloat.
Changed object positioning to follow Screen bounds.
- Fixed bug with Background View not scaling to the screen properly

Fixed "snapping" of Cloud View between animation cycles
Disabled autorotate.

Commit:
Merge branch 'dev_1.0' of https://github.com/cgosingtian/iOS-Exercises-2-Air-Raid into dev_1.0

# Conflicts:
#	README.md

Commit:
Info.plist - Corrected signing issues
